[2단계 - 출석 다시 구현하기] 훌라(김성준) 미션 제출합니다.

Sung-june27 opened the pull request:
<!-- 

## 코드 리뷰 팁

- 코드와 관련된 질문이 있다면, PR 본문에 적기 보다는 해당 코드를 선택하고 코멘트를 남겨주세요.
  - [참고: Adding comments to a pull request](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/commenting-on-a-pull-request#adding-comments-to-a-pull-request)

-->

## 체크 리스트

- [x] 미션의 필수 요구사항을 모두 구현했나요?
- [x] Gradle `test`를 실행했을 때, 모든 테스트가 정상적으로 통과했나요?
- [x] 애플리케이션이 정상적으로 실행되나요?
- [x] [프롤로그](https://prolog.techcourse.co.kr)에 셀프 체크를 작성했나요?
  - [프롤로그](https://prolog.techcourse.co.kr/studylogs/3951)


## 객체지향 생활체조 요구사항을 얼마나 잘 충족했다고 생각하시나요?

### 1~5점 중에서 선택해주세요.

- [ ] 1 (전혀 충족하지 못함)
- [ ] 2
- [ ] 3 (보통)
- [x] 4
- [ ] 5 (완벽하게 충족)

### 선택한 점수의 이유를 적어주세요.

최대한 모든 원시값을 포장하려고 노력했지만,
단순 상수로만 빼낸 경우도 존재했고,
AttendanceBook의 필드인 Map에서 키값을 String 그대로 사용했습니다.
그러나, 이 경우 String을 포장할 이유를 아직 느끼지 못해서 포장하지 않았습니다.

## 어떤 부분에 집중하여 리뷰해야 할까요?

안녕하세요 샐리!
지난 step1 때 빨리 merge 시켜주셔서 step2를 빨리 시작할 수 있었습니다!
이번 Step2는 TDD를 지키기 위해서 노력해봤어요.
때문에... step2는 오래 걸린 것 같아요 😂

이번 리뷰도 잘 부탁드립니다!

### 적용한 것들
- TDD를 적용해봤습니다.
  - domain의 테스트 커버리지를 확인해봤을 때, getter 같은 메서드를 제외하고 테스트가 잘 적용되는 것을 보았을 때, 성공적이라고 생각합니다.
  - 그러나, TDD를 적용하는 과정에서 객체를 어느 시점에 어떻게 분리를 할 것인가에 대해 생각하느라 시간이 오래 걸렸습니다.
  - TDD를 적용하니, 리팩토링을 하며 버그가 발생하는 것에 대한 두려움이 없어서 좋았던 것 같습니다.
  - 그러나... 아직 TDD 과정에서 객체를 분리하는 것에 대해서는 많이 어려웠습니다.
  - 또, controller를 작성하며 반환값/파라미터 타입을 많이 바꾼 것 같습니다.

- 테스트 클래스 내에서 팩토리 메서드를 이용해 생성해주는 방식을 적용해봤습니다.
  - 테스트 코드에서 도메인을 생성하는 부분을 계속해서 복사 붙여넣기 하는 것이 부담되었는데, 해당 방법을 적용해보니 괜찮았던 것 같습니다. (실수 방지)
  - ```java
    public static AttendanceBook generateAttendanceBook(String nickname, List<LocalDateTime> dateTimes) {
        Map<String, Attendances> crewAttendances = Map.of(nickname, generateAttendances(dateTimes));
        return new AttendanceBook(new HashMap<>(crewAttendances));
    }

- null-safe하게 equals()를 비교하도록 적용
  - equals()의 순서를 변경하는 것만으로도 null-safe하게 비교해줄 수 있다는 것을 리뷰를 보고 알게 되었습니다! 감사합니다!!
  - ```java 
        if (Menu.ATTEND.equals(selectedMenu)) {
            attend();
        }
        if (Menu.UPDATE_ATTENDANCE.equals(selectedMenu)) {
            updateAttendance();
        }
        if (Menu.PRINT_ATTENDANCES_BY_CREW.equals(selectedMenu)) {
            printAttendancesByCrew();
        }
        if (Menu.PRINT_WARNING.equals(selectedMenu)) {
            printWarningCrews();
        }
        if (Menu.QUIT.equals(selectedMenu)) {
            System.exit(0);
        }
      ```

- `AttendanceBook`에서 `Map<String, Attendances> crewAttendances`형태로 데이터 유지
  - 이 프로그램에서는 단순히 닉네임을 표현하는 것 외에 검증과 같은 역할이 없다고 생각했기 때문에 포장하지 않았습니다.
  - 추후 유지보수가 필요하다면, 그 때 가서 감싸는 게 맞다고 생각했습니다.

- Enum을 활용해봤습니다.
 - 지난 ErrorMessage 피드백을 반영해, 프로그램 기능에 대한 에러는 통합해서 관리해봤습니다.
 - 또, 관련 있는 상수들을 모아서 관리해봤습니다.

### 고민
초기화하는 부분에 대해서 고민을 많이 했던 것 같습니다.
- 지난 피드백인 [`다만, 컨트롤러 영역에서 파일을 읽고, 파싱하고, 포맷을 맞추는게 좋은 구조일지 생각해보면 좋을 것 같아요 😄`](https://github.com/woowacourse/java-attendance/pull/61#discussion_r1966463931)에 대해서 많이 고민해봤던 것 같습니다.
- 또, 하나의 객체로 두자니, 단순 기능만 실행하는 것 같아 util로 관리했습니다.
- 컨트롤러의 영역이라고는 생각할 수 없겠구나 싶어서 attendance.util로 위치를 옮겼으나, 전체적인 구조 자체에 대해서 아직 어떤 방식이 더 좋을지 생각해보았지만 잘 떠오르지 않았던 것 같습니다...

Sung-june27 committed:
docs: 기능 요구 사항 초안 작성

Sung-june27 committed:
feat: 주말 및 공휴일에는 출석 불가한 기능 구현

Sung-june27 committed:
feat: 캠퍼스 운영 시간 외에는 출석 불가한 기능 구현

Sung-june27 committed:
feat: 해당 요일의 시작 시각으로부터 지각/결석/출석 판별 구현

Sung-june27 committed:
feat: 닉네임과 시간을 입력하면 출석하는 기능 구현

Sung-june27 committed:
feat: 닉네임과 시간을 입력해 출석하는 기능 구현

Sung-june27 committed:
refactor: Attendances 객체 분리

Sung-june27 committed:
refactor: Attendances 객체 AttendanceDate/AttendanceTime Map 적용

Sung-june27 committed:
refactor: Attendance 객체 적용

Sung-june27 committed:
test: 테스트용 팩토리 메서드 패턴 적용

Sung-june27 committed:
feat: 닉네임과 시간을 이용해 출석을 수정한다

Sung-june27 committed:
feat: 출결 횟수를 계산한다

Sung-june27 committed:
feat: 출석 기록을 바탕으로 제적 위험자를 파악하는 기능 구현

Sung-june27 committed:
refactor: 계층 분리

Sung-june27 committed:
feat: 파일 입출력을 통해 파일 읽기 구현

Sung-june27 committed:
feat: 파일 입력을 가공하는 기능 구현

Sung-june27 committed:
feat: AttendanceBook 객체 생성 팩토리 클래스 구현

Sung-june27 committed:
feat: 컨트롤러 메뉴 선택 기능 구현

Sung-june27 committed:
feat: 출석 기능 구현

Sung-june27 committed:
feat: 닉네임과 날짜를 이용해 출석 정보를 가져오는 기능 구현

Sung-june27 committed:
feat: 출석 수정 기능 구현

Sung-june27 committed:
feat: 닉네임을 통해 출석 조회하는 기능 구현

Sung-june27 committed:
feat: 위험자 정렬 및 출력 기능 구현

Sung-june27 committed:
refactor: 출석 기능 리팩토링

- 상수화
- 파라미터 final 키워드 적용

Sung-june27 committed:
refactor: 출석 수정 기능 리팩토링

-메서드명
-final 키워드 적용

Sung-june27 committed:
refactor: 출석 조회 기능 리팩토링

-변수명
-final 키워드 적용

Sung-june27 committed:
refactor: 위험자 조회 기능 리팩토링

-반복문에 stream 적용
-제적/면담/경고 로직 이동
-final 키워드 적용

Sung-june27 committed:
refactor: 컨트롤러 리팩토링

Sung-june27 committed:
refactor: 메서드명 및 패키지 이동

Sung-june27 committed:
refactor: Warning 클래스명 변경

Sung-june27 committed:
test: 테스트 코드 리팩토링 및 enum 상수 적용

Sung-june27 committed:
refactor: indent 및 메서드 라인, 패키지 위치 리팩토링

Sung-june27 committed:
docs: 기능 요구 사항 체크리스트 작성

Sung-june27 committed:
Merge branch 'sung-june27' into step2

# Conflicts:
#	README.md
#	src/main/java/attendance/Application.java
#	src/main/java/attendance/constant/ErrorMessage.java
#	src/main/java/attendance/controller/AttendanceController.java
#	src/main/java/attendance/domain/Attendance.java
#	src/main/java/attendance/domain/AttendanceStatus.java
#	src/main/java/attendance/domain/Attendances.java
#	src/main/java/attendance/domain/Menu.java
#	src/main/java/attendance/dto/AttendanceResultResponse.java
#	src/main/java/attendance/view/InputView.java
#	src/main/java/attendance/view/OutputView.java
#	src/test/java/attendance/domain/AttendanceTest.java

Sung-june27 committed:
conflict resolve

dusdn1702 commented on src/main/java/attendance/controller/AttendanceController.java:
```diff
@@ -1,113 +1,109 @@
-        while (!processMenu(today, crews)) {
+        attendanceBook = AttendanceBookFactory.create(
+                CrewAttendancesDataParser.parse(AttendancesFileReader.read()), today);
+        while (!processMenu()) {
```
empty body 로 돌리는게 어색하게 느껴지네요~ 다른 방법을 찾아볼까요?

dusdn1702 commented on src/main/java/attendance/controller/AttendanceController.java:
```diff
@@ -1,113 +1,109 @@
-        Crews crews = crewsService.init(CrewAttendanceParser.parseCrewAttendances(AttendancesFileReader.read()), today);
-        while (!processMenu(today, crews)) {
+        attendanceBook = AttendanceBookFactory.create(
+                CrewAttendancesDataParser.parse(AttendancesFileReader.read()), today);
```
```suggestion
String crewAttendancesData = CrewAttendancesDataParser.parse(AttendancesFileReader.read();
attendanceBook = AttendanceBookFactory.create(crewAttendancesData), today);
```
이렇게 바꾸면 더 읽기 편하겠네요!

dusdn1702 commented on src/main/java/attendance/controller/util/DateTimeConverter.java:
```diff
@@ -0,0 +1,42 @@
+
+public class DateTimeConverter {
+    private static final String TIME_FORMAT = "HH:mm";
+    private static final int START_DAY_OF_YEAR = 1;
```
년도의 시작일인가요 ? 🤔

Sung-june27 replied:
바로 수정했습니다....!!!😅

dusdn1702 commented on src/main/java/attendance/controller/util/DateTimeConverter.java:
```diff
@@ -0,0 +1,42 @@
+        try {
+            return LocalTime.parse(input, formatter);
+        } catch (DateTimeParseException e) {
+            throw new IllegalArgumentException("[ERROR] 시간을 올바른 형식으로 입력해주십시오.");
```
ErrorMessage로 분리하는 기준은 무엇인가요?

Sung-june27 replied:
`약속된 에러메시지인가?`를 기준으로 세우고 분리했습니다.
첫 번째로, 요구사항/실행 결과 예시에 있다면 약속된 에러메시지라고 간주했습니다.
추가적으로, 핵심적인 예외사항에 대해서도 enum으로 통합해 관리했습니다.

약속된 에러메시지 외에는 다른 개발자가 굳이 알 필요가 없다고 생각했기 때문에,
일반적인 에러메시지는 하드코딩했습니다!!

dusdn1702 commented on src/main/java/attendance/domain/Attendance.java:
```diff
@@ -1,77 +1,59 @@
 
     public static Attendance from(final LocalDateTime dateTime) {
-        return new Attendance(dateTime);
+        return new Attendance(AttendanceDate.from(LocalDate.from(dateTime)),
```
이렇게 of 와 from 으로 나누어 팩토리 메서드를 만든 이유가 무엇인가요? 
from이라는 단어가 세줄에 5개가 나오다니.. 너무 은닉화되어 있지는 않은지 고민해봐요!

Sung-june27 replied:
LocalDateTime 자체로도 Attendance 객체를 생성하고자 해서 나누어서 만들었어요.

`은닉화`에 대해서는 여지껏 생각해보지 못한 것 같은데, 이번 기회에 찾아보고 생각하게 되었어요!!
앞으로는 다른 개발자가 그 메서드 내부 구현을 쉽게 이해하지 못할 수 있다면, 이를 해결하는 방법을 강구해야하겠네요.
때문에 이번 리팩토링에서는 from을 지우고, 외부에서 AttendanceDate/AttendanceTime을 생성하는 방식으로  모두 통일했습니다.

[refactor: Attendance from 메서드 삭제](https://github.com/woowacourse/java-attendance/pull/93/commits/234c5101bc484c89d6590f0e4fca20051567a09f)

❓그런데, 하나 궁금한 점이 생겼어요!
은닉화의 '메서드 내부 구현을 쉽게 이해하지 못한다.'라는 부분에서,
그렇다면 from이 아닌 메서드명을 `createByLocalDateTime`과 같은 식으로 지어준다면
어느정도 내부 구현이 아니더라도 파라미터와 실행 결과가 예측될 수 있기 때문에,
이 또한 과도한 은닉화를 해결하는 방법이라고 볼 수 있을까요?

dusdn1702 replied:
흠 네 그렇게 해결할 수도 있겠지만, 
AttendanceDate/AttendanceTime 이 꼭 필요한 객체인지에 대해서도 생각해보면 좋을 것 같아요. 
예를 들어 AttendanceTime의 경우, 두가지 메서드가 존재해요.
- 출석 상태 확인
- 시간 검증
근데 이 두 메서드 각각 AttendanceStatus, CampusOperatingRule의 역할이라고 볼 수 있죠.
이런 경우 이 AttendanceTime을 생성하는게 맞을까요? 
그렇게 했을 때의 장점이 무엇일까요?

Sung-june27 replied:
우선, 샐리의 말대로 AttendanceDate와 AttendanceTime을 사용하지 않는 방향으로 한 번 구현해봤어요.

이전에는 enum을 상수를 저장하는 용도로만 사용을 했었는데,
각각 역할을 부여해주게 되니 AttendanceDate와 AttendanceTime을 사용하지 않아도 되는 것 같아요.

AttendanceTime을 생성했을 때의 장점이라기보다는 선택이라고 생각이 들었어요.
해당 역할을 AttendanceTime 객체에 부여를 해주느냐, 아니면 enum 객체에 부여를 해주느냐에 따라 달라지는 것 같았어요.
저는 그 역할을 객체에게 부여해주기로 선택해서 생성을 했던 것이고요!

그러나, 객체를 생성했을 때의 단점은 객체 그래프가 한 단계 더 깊어지는 것이라고 생각해요.

`원시값을 포장한다.`라는 요구사항에 맞춰서 Date/Time을 각각 포장했었는데,
trade-off를 고려해서 잘 선택해야겠다고 생각이 들었네요.
이 때, enum에서 처리하는 방법도 한 번 고려해볼만 하다는 생각을 하는 계기가 되었어요!

dusdn1702 replied:
맞아요, 훌라 말처럼 선택의 영역이라고 저도 생각해요.
다만 너무 모든 원시값을 포장하려고 하다보면 이렇게 복잡도가 올라가는 경우가 생겨서 이 부분에 대해서도 한번 생각해보면 좋을 것 같아서 말씀드렸어요 👍

dusdn1702 commented on src/main/java/attendance/domain/Attendance.java:
```diff
@@ -1,77 +1,59 @@
-        }
-        return AttendanceStatus.ATTEND;
+    public AttendanceStatus checkAttendanceStatus() {
+        return attendanceTime.checkAttendanceStatus(attendanceDate.isMonday());
```
출석 상태를 확인하는데 출석시간에 가서 다시 출석 상태를 확인하고, 그때 넘기는 값은 출석 일자의 월요일 여부라니 코드 이해가 조금 어렵네요.
더 잘 풀어낼 수 있을 것 같아요.

Sung-june27 replied:
AttendanceDate/AttendanceTime 간의 의존성을 끊어주기 위해
Attendance 객체에서 absenceThreshold/lateThreshold를 구해서 넘겨주는 방식으로 리팩토링해봤어요!

[refactor: 출결 판단 메서드 리팩터링](https://github.com/woowacourse/java-attendance/pull/93/commits/076389637c9797e47729d75327e5987ab1cd5ae5)

dusdn1702 commented on src/main/java/attendance/domain/Attendance.java:
```diff
@@ -1,77 +1,59 @@
 
-    public boolean isEqualToDate(final LocalDate today) {
-        return today.equals(LocalDate.from(dateTime));
+    public AttendanceDate getAttendanceDate() {
```
getter를 사용하지 않는 방향으로 리팩터링해보면 좋겠어요.

Sung-june27 replied:
```java
    public LocalDateTime getAttendanceDateTime() {
        return LocalDateTime.of(this.attendanceDate.date(), this.attendanceTime.time());
    }
```
이와 같은 방식으로 바꿔보았어요
객체 자체를 넘겨주는 것이 아닌, LocalDateTime을 Attendance 객체에서 만들어서 넘겨주는 방식으로요!

하지만 아직 메서드명이 get~인 것은 저도 마음에 들지는 않네요...😭
혹시 샐리는 이런 경우 어떤 식으로 네이밍하시나요?

dusdn1702 replied:
get~ 이 맞는 상황이네요 ㅎㅎ 저도 동일하게 네이밍했을 것 같아요!

dusdn1702 commented on src/main/java/attendance/domain/Attendance.java:
```diff
@@ -1,77 +1,59 @@
-    public AttendanceStatus getStatus() {
-        return status;
+    @Override
+    public boolean equals(Object o) {
```
오버라이딩이 필요했던 이유가 있을까요?

Sung-june27 replied:
네오가 강의 중 'equals()를 오버라이딩해서 사용한다면, getter를 사용하지 않고도 두 객체를 비교할 수 있다.'라고 했던 것이 기억이 나서 사용해봤어요.
때문에, 테스트 단계에서 우선적으로 필요해져서 오버라이딩을 했고, 추후 설계에 따라 충분히 프로덕션 코드에서도 사용될 수 있다고 생각했어요. (assertj isEqualTo)
그러나, 사용되지 않았네요.... 바로 삭제했습니다!

dusdn1702 commented on src/main/java/attendance/domain/Attendances.java:
```diff
@@ -1,73 +1,69 @@
-    public void addAttendance(Attendance attendance) {
+    public Attendance add(final Attendance attendance) {
+        if (existsByDate(attendance)) {
+            throw new IllegalArgumentException(ALREADY_ATTEND.getMessage());
```
add라는 함수명으로 출석을 추가하면서 이미 존재하면 추가되지 않을 것이라고 기대할 수 있을까요? 함수 분리 혹은 함수명 변경이 좋겠어요!

Sung-june27 replied:
addIfAbsent로 메서드명 변경!

Map에 computeIfAbsent라는 메서드가 있더라고요
메서드명을 짓기 어려울 때는 자바 api의 네이밍을 참고하는 것도 하나의 방법이 된다고 들어서,
한 번 참고해보았습니다!

dusdn1702 replied:
좋네요 ㅎㅎ 💯

dusdn1702 commented on src/main/java/attendance/domain/AttendanceBook.java:
```diff
@@ -0,0 +1,53 @@
+    public Attendance updateAttendance(final String nickname, final LocalDateTime dateTime) {
+        Attendances attendances = crewAttendances.get(nickname);
+        Attendance attendance = attendances.updateAttendance(dateTime);
+        crewAttendances.put(nickname, attendances);
```
AttendanceBook에서 attendance를 수정하는데 cresAttendances에서 attendances를 get 하고 put하는게 이상하게 느껴지네요.
변수명을 더 이해하게 쉽게 변경하면 좋겠어요.

Sung-june27 replied:
아직 깊은 복사/얕은 복사에 대해서 이해도가 낮아서 이런 코드가 나온 것 같아요
코드를 이리저리 다시 구현해보니, put하지 않아도 수정은 정상적으로 이루어지더라고요!
때문에 get해서 update만 해주는 방식으로 변경했습니다!

아마... 변수명을 바꿨다면 searchedAttendances와 같은 것으로 바꿔봤을 것 같네요

[refactor: AttendanceBook 리팩토링](https://github.com/woowacourse/java-attendance/pull/93/commits/a354826b6f81e40e63ef487184aa86844c8fcd81)

dusdn1702 commented on src/main/java/attendance/domain/AttendanceBook.java:
```diff
@@ -0,0 +1,53 @@
+        return attendance;
+    }
+
+    public Attendances findByNickname(final String nickname) {
```
위 함수들에서 이 함수를 사용하지 않은 이유가 있나요?

Sung-june27 replied:
처음에 고민해보았을 때는 'map.get(nickname) 정말 간단한 코드인데, 굳이 findByNickname이라는 메서드를 사용할 필요가 있을까?'라는 생각을 했어요.
왜냐하면, '최소한의 기능 구현을 한다.'라는 점에 더 꽂혀있었던 것 같네요...

이번에 피드백을 반영하며 다시 곰곰히 생각해보니,
TDD의 3번째 단계인 '리팩토링' 단계에서 합쳐서 코드의 중복을 제거하는 것이 더 좋은 방향이었던 것 같습니다!
때문에 이번에 다시 사용하는 방향으로 리팩토링했어요!

dusdn1702 commented on src/main/java/attendance/domain/AttendanceBook.java:
```diff
@@ -0,0 +1,53 @@
+    public Map<String, Attendances> findPenaltyCrews() {
+        return crewAttendances.entrySet()
+                .stream()
+                .filter(entry -> !AttendancePenalty.NONE.equals(entry.getValue().calculatePenalty()))
```
조금 더 가독성 좋게 바꿔보면 어떨까요?
이 함수를 이해하기 위해서 map에 어떤 값이 있는지, value는 무엇인지, calculate은 어떻게 동작하는지 확인해야하네요.

Sung-june27 replied:
```java
    private boolean hasPenalty(Entry<String, Attendances> entry) {
        return !AttendancePenalty.NONE.equals(entry.getValue().calculatePenalty());
    }
```

메서드 분리를 통해서 가독성을 조금이나마 올려봤어요!

dusdn1702 commented on src/main/java/attendance/domain/AttendanceBook.java:
```diff
@@ -0,0 +1,53 @@
+    private void validateNickname(final String nickname) {
+        if (!crewAttendances.containsKey(nickname)) {
+            throw new IllegalArgumentException(UNREGISTERED_NICKNAME.getMessage());
+        }
```
```suggestion
        if (crewAttendances.containsKey(nickname)) {
            return;
        }
        throw new IllegalArgumentException(UNREGISTERED_NICKNAME.getMessage());
        }
```
!를 이용한 반대 로직은 이해하기 어렵고 실수하기 쉬워 지양하는 편이에요~!

Sung-june27 replied:
부정 연산자도 다른 개발자가 봤을 때 그런 견해가 있을 수도 있겠군요!
다른 사람이 코드를 봤을 때를 더 고려해주어야겠어요

❓혹시, 부정 연산자를 조건문 내에서만 지양하시는 편인가요?
해당 피드백을 반영하며 이번에 리팩토링을 하다보니, controller의 executeMenu() 내에서
```java
return !Menu.QUIT.equals(selectedMenu);
```
이렇게 구현을 하게 되더라고요.
샐리는 혹시 반환값에 대해서도 부정연산자를 지양하시는 편인가요? 아니면 조건문 내에서만 지양하시는 편인가요?

dusdn1702 replied:
부정연산자는 개발자로 하여금 한번 더 반대로 생각하도록 하는 코드라 모든 코드에서 지양하는 편이에요! 사용하지 않고도 구현할 수 있거든요.

`return !Menu.QUIT.equals(selectedMenu);` -> QUIT과 동일하지 않은 값
`return Menu.QUIT.equals(selectedMenu);` -> QUIT

dusdn1702 commented on src/main/java/attendance/domain/AttendanceDate.java:
```diff
@@ -0,0 +1,38 @@
+        return this.date.isEqual(date);
+    }
+
+    private static void validate(final LocalDate date) {
```
어떤 유효성 검증인지 메서드명을 자세히 적어주세요.
validate은 공휴일이 걸러질 것이라 기대되지 않아요.

Sung-june27 replied:
`validateAvailableAttendDate` 메서드명을 좀 더 명확하게 나타내는 방식으로 변경해봤습니다!

dusdn1702 commented on src/main/java/attendance/domain/AttendancePenalty.java:
```diff
@@ -0,0 +1,33 @@
+public enum AttendancePenalty {
+    EXPULSION("제적", 5),
+    INTERVIEW("면담", 3),
+    Warning("경고", 2),
```
👀

Sung-june27 replied:
CAU....CAUTION으로 바꿨습니다...!!!!

클래스명 네이밍에 정신 팔려서 소홀했네요.....

dusdn1702 commented on src/main/java/attendance/domain/AttendancePenalty.java:
```diff
@@ -0,0 +1,33 @@
+    }
+
+    public static AttendancePenalty from(final int value) {
+        if (value > EXPULSION.criteria) {
+            return EXPULSION;
+        }
+        if (value >= INTERVIEW.criteria) {
+            return INTERVIEW;
+        }
+        if (value >= Warning.criteria) {
+            return Warning;
+        }
```
깔끔하네요 👍

dusdn1702 commented on src/main/java/attendance/domain/AttendancePenalty.java:
```diff
@@ -0,0 +1,33 @@
+package attendance.domain;
+
+public enum AttendancePenalty {
```
페널티 enum을 도메인 영역에 두신 이유가 궁금해요

Sung-june27 replied:
AttendancePenalty가 완전히 상수만을 위한 enum이 아닌, from 메서드에 로직이 포함되어있고,
또 프로그램에 핵심적인 데이터가 들어있다고 판단해 domain영역에 두었어요

dusdn1702 commented on src/main/java/attendance/controller/util/DateTimeConverter.java:
```diff
@@ -0,0 +1,42 @@
+import java.time.format.DateTimeFormatter;
+import java.time.format.DateTimeParseException;
+
+public class DateTimeConverter {
```
controller.util과 외부 util 패키지에 차이가 있나요?

Sung-june27 replied:
사실, 차이는 없다고 생각하지만 '사용되는 위치에 가깝게'라는 의도를 가지고 `controller.util` 패키지를 만들었습니다.
다만 '파일입출력'과 연관되었다는 것 때문에 'resources 폴더에 접근한다면, controller 계층에만 연관된 것은 아니지 않나?'라는 생각을 해서 `외부  util` 패키지를 만들게 되었습니다.

계층을 구분하는 저만의 기준을 세워야겠다고 생각 중인데, 아직은 모호해서 이런 고민을 하는 것 같습니다...😭

dusdn1702 replied:
아직은 프로그램의 규모가 그렇게 크지 않기 때문에, 본인의 기준을 가지고 구현하면 된다고 생각해요.
지금처럼 훌라만의 기준을 세워가는 고민을 계속 하면 좋을 것 같아요 👍

dusdn1702 commented on src/main/java/attendance/domain/Attendances.java:
```diff
@@ -1,73 +1,69 @@
-                    return status.equals(AttendanceStatus.ABSENCE) || status.equals(AttendanceStatus.LATE_ABSENCE);
-                })
-                .count();
+    public AttendancePenalty calculatePenalty() {
+        int totalAbsenceCount = countAbsence() + countLate() / LATE_TO_ABSENCE_UNIT;
+        return AttendancePenalty.from(totalAbsenceCount);
```
깔끔하고 좋네요 💯

dusdn1702 commented on src/main/java/attendance/dto/AttendanceResultResponse.java:
```diff
@@ -2,13 +2,17 @@
+    public static AttendanceResultResponse from(final Attendance attendance) {
+        LocalDate date = attendance.getAttendanceDate().date();
+        LocalTime time = attendance.getAttendanceTime().time();
+        return new AttendanceResultResponse(LocalDateTime.of(date, time), attendance.checkAttendanceStatus());
```
dto 에서 로직을 수행시키는 것이 좋은 방향일까요?

Sung-june27 replied:
엄격하게 본다면, DTO에게 로직을 수행시키면 안 되지만 '포맷팅 정도는 괜찮지 않을까?'라는 안일한 생각을 가진 것 같네요

이번 미션에서는 servie 계층을 사용하지 않으려고 하다보니 이런 코드가 나온 것 같아요.
`이 프로그램에서 정말 service라는 계층이 필요한가?`라는 생각을 하고 있었거든요!

dusdn1702 commented on src/main/java/attendance/dto/PenaltyCrewsResponse.java:
```diff
@@ -0,0 +1,35 @@
+        );
+    }
+
+    public record PenaltyCrew(
```
public 값이라 도메인에서 PenaltyCrew 라는 객체가 필요하게 되면 가져다 쓸 수 있겠어요! 위험하지 않을까요?

Sung-june27 replied:
View에서 PenaltyCrew객체의 정보를 출력하기 위해서는 public 접근제어자를 통해 허용해주어야 하더라고요..
대신, 클래스명에 Response를 추가해 dto라는 의미를 더 부여해주었습니다.

dusdn1702 commented on src/main/java/attendance/dto/AttendanceResultResponse.java:
```diff
@@ -2,13 +2,17 @@
 ) {
-    public static AttendanceResultResponse from(Attendance attendance) {
-        return new AttendanceResultResponse(attendance.getDateTime(), attendance.getStatus());
+    public static AttendanceResultResponse from(final Attendance attendance) {
```
dto는 어떤 목적이라고 생각하시나요? 노출(view)을 위한 값을 가지며 가장 핵심(domain) 데이터도 다 알고 있네요?

Sung-june27 replied:
제가 생각하는 DTO는
- 각 계층 간, 데이터를 감싸 이동시켜주는 컨테이너
- MODEL과 VIEW 간의 관계를 끊어주는 역할
이라고 생각해요!

때문에, 이 DTO에서 샐리가 말한대로 가장 핵심 데이터인 dateTime을 알고 있다고 하더라도,
Attendance 객체 자체를 View에서 사용하는 것보다는 더 안전하다고 생각해 사용했어요.

dusdn1702 commented on src/main/java/attendance/util/AttendanceBookFactory.java:
```diff
@@ -0,0 +1,56 @@
+
+    private static List<Attendance> initAttendances(final LocalDate today) {
+        List<Attendance> attendances = new ArrayList<>();
+        for (int day = 1; day < today.getDayOfMonth(); day++) {
+            try {
+                LocalDateTime dateTime = LocalDateTime.of(today.withDayOfMonth(day), DEFAULT_TIME);
```
depth를 줄여봐요!

dusdn1702 replied:
이부분 무슨 로직인가요??

Sung-june27 replied:
이 로직을 통해 모두 결석인 Attendances 객체를 만들어주고,
파일 내용에 따라 update해주는 방식으로 구현했어요

이 부분은 모두 결석인 Attendances 객체를 만들기 위해 DEFAULT_TIME으로 결석인 시간대(17:59)를 넣어주는 로직이에요!


+) depth는 줄여보았습니다!
[refactor: AttendanceBookFactory indent 줄이기](https://github.com/woowacourse/java-attendance/pull/93/commits/c46eaee7be8f49a98be93bef866072775110557e)

dusdn1702 replied:
DEFAULT_TIME 이 결석의 시간이라는걸 모두가 이해할 수 있을까요? 모든 코드를 읽은 저조차 가늠하지 못했어요!
출석 기준 시간이 18시로 변경되면 저는 이 부분을 놓치고 10시 값만 수정했을텐데, 이런 위험성을 줄일 수 있는 코드로 바꿔보면 어떨까요?

Sung-june27 replied:
```java
            LocalDate date = today.withDayOfMonth(day);
            LocalTime absenceTime = CAMPUS_CLOSE_HOUR.getTime().minusMinutes(1);
            return Attendance.from(date, absenceTime);
```

이 부분에서도 absenceTime을 enum을 통해 생성해주는 방식으로 한 번 변경해봤어요!

dusdn1702 commented on src/main/java/attendance/util/AttendanceBookFactory.java:
```diff
@@ -0,0 +1,56 @@
+        return attendanceBook;
+    }
+
+    private static void applyAttendances(Entry<String, List<LocalDateTime>> entry, AttendanceBook attendanceBook) {
```
```
nickname,datetime
쿠키,2024-12-13 10:08
빙봉,2024-12-13 10:07
빙티,2024-12-01 10:00
빙티,2024-12-02 10:00
빙티,2024-12-03 10:00
빙티,2024-12-04 10:00
빙티,2024-12-05 10:00
빙티,2024-12-06 10:00
빙티,2024-12-09 10:00
빙티,2024-12-10 10:00
빙티,2024-12-11 10:00
빙티,2024-12-12 10:00
빙티,2024-12-13 10:00
빙티,2024-12-16 10:07
이든,2024-12-13 10:07
```
이렇게 csv 파일을 수정했는데요~!
`[ERROR] 출석하지 않은 날짜입니다.` 라고 하네요 🤔 
저는 사용자로서 이유를 이해하지 못했어요 고쳐주세요! 😄

Sung-june27 replied:
샐리의 csv 파일 중, 빙티의 12월 1일 기록이 존재하는데 12월 1일은 주말이라서 출석할 수 없는 날짜에요!
12월 1일의 기록을 삭제한다면 잘 돌아가네요~!

이번 미션을 구현하며, csv 파일 자체는 온전한 출석 기록을 가지고 있다는 가정하에 진행했기 때문에 파일의 입력에 대해서는 다시 한 번 검증하는 로직이 존재하지 않아서 발생한 에러 같아요!

dusdn1702 replied:
그렇군요! 에러 메시지만 사용자가 이해하기 쉽게 고쳐져도 좋을 것 같아요 ㅎㅎ

dusdn1702 commented on src/main/java/attendance/util/AttendanceBookFactory.java:
```diff
@@ -0,0 +1,56 @@
+package attendance.util;
```
이 초기화 하는 클래스의 위치 고민을 말씀 주셨는데요. 오히려 controller.util에 가까워보여요. 가장 핵심 도메인인 출석부를 생성하는데 이 위치에 있으면 가져다 사용하기 너무 쉬워 위험할 것 같아요.

Sung-june27 replied:
`가져다 사용하기 너무 쉬워 위험할 것 같아요.`
이 말을 들으니 클래스에게 적절한 위치를 부여하는 것도 중요하구나 생각이 드네요!
이전까지는 무의식적으로 뭉뚱그려서 관리했는데, 이러한 관점에서도 봐야겠구나 생각이 들었어요

dusdn1702 commented on src/test/java/attendance/domain/AttendanceBookTest.java:
```diff
@@ -0,0 +1,116 @@
+        assertThat(result.keySet()).hasSize(2);
+    }
+
+    public static AttendanceBook generateAttendanceBook(String nickname, List<LocalDateTime> dateTimes) {
```
```suggestion
    private static AttendanceBook generateAttendanceBook(String nickname, List<LocalDateTime> dateTimes) {
```

dusdn1702 commented on src/test/java/attendance/domain/AttendanceBookTest.java:
```diff
@@ -0,0 +1,116 @@
+import java.util.Map;
+import org.junit.jupiter.api.Test;
+
+public class AttendanceBookTest {
+
+    @Test
+    void 출석_시_닉네임이_존재하지_않으면_예외를_던진다() {
```
테스트들이 좋네요 👍

dusdn1702 commented on src/test/java/attendance/domain/AttendanceBookTest.java:
```diff
@@ -0,0 +1,116 @@
+        List<LocalDateTime> attendances = List.of(LocalDateTime.of(2024, 12, 13, 9, 59));
+        AttendanceBook attendanceBook = generateAttendanceBook("훌라", attendances);
+
+        final var result = attendanceBook.updateAttendance("훌라", LocalDateTime.of(2024, 12, 13, 10, 6));
```
var의 이유가 있나요?

Sung-june27 replied:
강의에서 네오가 테스트 코드에서 final var을 통해 결과값을 받으며, '반환값이 중요하지 않아서.'라는 말을 했습니다.

그래서 한 번 이번에 적용시켜봤어요! 저 또한, 반환값을 너무 상세하게 생각하려고 하다보니 TDD 단계에서 구조에 대한 너무 먼 설계까지 하려고 하는 모습을 보고 한 번 적용시켜 봤습니다.

그런데, 반환값에 대해 깊게 생각하지 않다보니 TDD의 흐름 자체는 매끄러워졌지만, 나중에 설계 부분에서 좀 많이 바꾸어주는 경우가 있었어요. 이 방법은 장단점이 확실하다고 생각이 드네요

dusdn1702 commented on src/test/java/attendance/domain/AttendanceTest.java:
```diff
@@ -1,82 +1,94 @@
-        Attendance attendance = Attendance.from(LocalDateTime.of(2024, 12, 16, 14, 0, 0));
-        assertThat(attendance.isEqualToDate(LocalDate.of(2024, 12, 16))).isTrue();
+    void 다른_요일_출결_상태를_반환한다_지각() {
+        Attendance attend = Attendance.from(LocalDateTime.of(2024, 12, 13, 10, 6));
```
출석 시간이 12시로 바뀌어도 로직은 변하지 않을텐데요.
여기서도 AttendanceTimeTest 처럼 프로덕션 코드의 enum을 사용해 테스트가 실패해서 테스트를 변경해야하는 수고로움을 줄일 수 있을 것 같아요.

Sung-june27 replied:
적용해봤어요!

[test: AttendanceTest enum 적용](https://github.com/woowacourse/java-attendance/pull/93/commits/72b0e0090d1e4ffcb295439059d0765e261ed87e)

dusdn1702 commented on src/test/java/attendance/domain/AttendanceTimeTest.java:
```diff
@@ -0,0 +1,84 @@
+
+    @Test
+    void _08시_이전에는_출석할_수_없다() {
+        LocalTime time = CAMPUS_OPEN_HOUR.getTime().minusMinutes(1);
```
👍👍👍👍👍

dusdn1702 reviewed (changes requested):
안녕하세요 훌라 ~

전체적으로 도메인 객체를 잘 분리해서 생성하고, 각 객체에 행위를 부여하는 식으로 코드 작성해주셔서 이해하기 쉬웠고, 객체 분리가 되니 자연스레 TDD도 잘 수행해주신게 느껴졌어요. 고생 많으셨습니다. 👏 

몇가지 궁금한점 피드백 남겨 두었어요. 확인 부탁드릴게요!
궁금한 점 편하게 코멘트 남겨주세요 🙇

Sung-june27 committed:
refactor: Attendance getter 제거 및 equals 오버라이딩 삭제

Sung-june27 committed:
refactor: 변수 추출 및 while empty body 리팩토링

Sung-june27 committed:
refactor: 변수명 오류 수정

Sung-june27 committed:
refactor: AttendanceBook 리팩토링

- findByNickname() 메서드를 이용한 중복 코드 제거
- hasPenalty() 메서드 분리
- !(부정연산자)를 이용한 비교 방식 변경

Sung-june27 committed:
refactor: 네이밍 변경

Sung-june27 committed:
test: AttendanceTest enum 적용

Sung-june27 committed:
refactor: Attendance from 메서드 삭제

Sung-june27 committed:
refactor: AttendanceBookFactory indent 줄이기

Sung-june27 committed:
docs: attendances.csv 파일 수정

Sung-june27 committed:
refactor: util 패키지 경로 이동

Sung-june27 committed:
refactor: 출결 판단 메서드 리팩터링

Sung-june27 reviewed (commented):
샐리!! 정말 빠른 리뷰 감사해요

이번 피드백을 수행하며,
TDD와 함께 객체 설계는 전보다 많이 나아진 것 같지만
아직 메서드명/변수명을 통해 코드의 의도를 나타내는 것은 부족하다고 생각했어요.
이 부분은 좀 더 고쳐나가야겠네요...

이번 미션에서는 특히 test fixture와 객체 구조 설계에 대해서 많이 배운 것 같아요
그리고... 아직 계층에 대한 명확한 이해가 부족하다고 다시 한 번 생각이 드네요(특히 util, dto)
그래서 알맞지 않은 역할을 부여한 부분들도 피드백을 받으며 많이 발견하게 된 것 같아요

dusdn1702 commented on src/main/java/attendance/domain/AttendanceBook.java:
```diff
@@ -0,0 +1,51 @@
+    }
+
+    public void attend(final String nickname, final Attendance attendance) {
+        validateNickname(nickname);
+        findByNickname(nickname).addIfAbsent(attendance);
```
validateNickname 이 2번 수정되겠네요

dusdn1702 reviewed (changes requested):
안녕하세요 훌라!

피드백들 잘 반영해주셨네요 👍 
이전에도 잘 구현해주셔서 이번엔 코멘트 답변을 몇가지 남겨두었는데요~
더 개선할 여지가 있을 것 같아 한번 고민해보고 다시 요청 주세요 😄

Sung-june27 committed:
refactor: validateNickname 2번 호출되는 부분 삭제

Sung-june27 committed:
refactor: AttendanceBookFactory 상수 제거

- 상수를 제거하고, CampusOperatingRule 사용

Sung-june27 committed:
refactor: controller 부정연산자 제거

Sung-june27 committed:
refactor: AttendanceDate/Time 삭제

Sung-june27 committed:
refactor: ErrorMessage 수정

Sung-june27 committed:
refactor: final 키워드 적용

Sung-june27 committed:
refactor: LocalDate/LocalTime 분리

Sung-june27 reviewed (commented):
이번에 피드백을 받으며,
TDD 과정에서 RED-GREEN을 지키려고는 노력했지만, 리팩토링 단계에는 소홀했구나 느꼈어요.
또, '이정도면 되겠지'에서 한 번 더 들어가 다른 사람이 코드를 봤을 때를 더 고려해봐야겠다고 생각하는 계기가 되었습니다!

dusdn1702 reviewed (approved):
안녕하세요 훌라~
리뷰 반영해주신 부분들 확인했습니다! 처음 구현부터 깔끔한 구조로 같이 다양한 생각을 하도록 구현해주셔서 저도 많이 배웠어요 ㅎㅎ
이번 미션은 여기서 마무리하도록 할게요. 고생 많으셨습니다 😄
